internal/app/singularity/plugin_compile_linux.go

```diff
@@ -87,69 +108,59 @@ func CompilePlugin(sourceDir, destSif, buildTags string) error {
 // the plugin. buildPlugin returns the path to the built file, along with an error.
 //
 // This function essentially runs the `go build -buildmode=plugin [...]` command.
-func buildPlugin(sourceDir, buildTags string) (string, error) {
-	workpath, err := getSingularitySrcDir()
-	if err != nil {
-		return "", errors.New("singularity source directory not found")
-	}
-
+func buildPlugin(sourceDir string, bTool buildToolchain) (string, error) {
 	// assuming that sourceDir is within trimpath for now
 	out := pluginObjPath(sourceDir)
 
-	goTool, err := exec.LookPath("go")
-	if err != nil {
-		return "", errors.New("go compiler not found")
-	}
-
 	args := []string{
 		"build",
 		"-o", out,
 		"-trimpath",
 		"-buildmode=plugin",
-		"-tags", buildTags,
+		"-tags", bTool.buildTags,
 		sourceDir,
 	}
 
-	sylog.Debugf("Running: %s %s", goTool, strings.Join(args, " "))
+	sylog.Debugf("Running: %s %s", bTool.goPath, strings.Join(args, " "))
 
-	buildcmd := exec.Command(goTool, args...)
+	buildcmd := exec.Command(bTool.goPath, args...)
 
-	buildcmd.Dir = workpath
+	buildcmd.Dir = bTool.workPath
 	buildcmd.Stderr = os.Stderr
 	buildcmd.Stdout = os.Stdout
 	buildcmd.Stdin = os.Stdin
-	buildcmd.Env = append(os.Environ(), "GO111MODULE=on")
+	buildcmd.Env = bTool.envs
 
 	return out, buildcmd.Run()
 }
 
 // generateManifest takes the path to the plugin source, sourceDir, and generates
 // its corresponding manifest file.
-func generateManifest(sourceDir string) (string, error) {
+func generateManifest(sourceDir string, bTool buildToolchain) (string, error) {
 	in := pluginObjPath(sourceDir)
 	out := pluginManifestPath(sourceDir)
 
-	pluginPointer, err := plugin.Open(in)
-	if err != nil {
-		return "", err
+	args := []string{
+		"run",
+		"-trimpath",
+		"-tags", bTool.buildTags,
+		"cmd/plugin/manifest.go",
+		in,
 	}
 
-	sym, err := pluginPointer.Lookup(pluginapi.PluginSymbol)
-	if err != nil {
-		return "", err
-	}
+	manifest := &bytes.Buffer{}
 
-	p, ok := sym.(*pluginapi.Plugin)
-	if !ok {
-		return "", fmt.Errorf(`symbol "Plugin" not of type Plugin`)
-	}
+	runCmd := exec.Command(bTool.goPath, args...)
+	runCmd.Dir = bTool.workPath
+	runCmd.Stdout = manifest
+	runCmd.Stderr = os.Stderr
+	runCmd.Env = bTool.envs
 
-	manifest, err := json.Marshal(p.Manifest)
-	if err != nil {
-		return "", fmt.Errorf("could not marshal manifest to json: %v", err)
+	if err := runCmd.Run(); err != nil {
+		return "", fmt.Errorf("could not get manifest: %s", err)
 	}
 
-	if err := ioutil.WriteFile(out, manifest, 0644); err != nil {
+	if err := ioutil.WriteFile(out, manifest.Bytes(), 0644); err != nil {
```

Comment:
you could simply pass a second argument to the tool with the destination filename.

Reply:
Not sure what do you mean here. Could you explain a bit more? WriteFile takes bytes, not reader.

Reply:
I mean you could have the program write a file instead of spitting the manifest to stdout. If you do that, you don't need this at all, you just need to run the program, verify the exit code and return the filename, as you are doing now.

If you need to support writing the manifest to stdout, something line `manifest-gen input /dev/stdout` always works, or you could follow the convention that `manifest input -` means "write to stdout", or you could make the second argument optional, and fall back to stdout if it's not provided.

Reply:
got it, will update. Thanks